### PR TITLE
UI Central Dogma and A High-Level Compositional Design

### PR DESCRIPTION
**Pull request by alice-i-cecile**
[**RENDERED**](https://github.com/alice-i-cecile/rfcs/blob/ui-central-dogma/rfcs/ui-central-dogma.md)

This meta-RFC lays out a vision for work to be done to achieve a complete solution for `bevy_ui`. By breaking the problem down into a central dataflow and many competing UI-like behaviors, we can create a modular design that can be designed, implemented and replaced in smaller chunks and reused in flexible ways for designs that straddle gameplay and "classical UI".

This is intended to be an extremely collaborative process: please comment and make [pull requests](https://github.com/alice-i-cecile/rfcs/pulls) aggressively.

This RFC contains several **benchmark paper prototypes**, intended to be designed and completed by community members to share their visions of what an ideal UI looks like. Feel free to submit yours in the appropriate section (in a [collapsible section](https://gist.github.com/pierrejoubert73/902cc94d79424356a8d20be2b382e1ab)) regardless of how many you've completed or the status of your prototypes / RFCs.

P.S. This RFC uses the proposed lightly tweaked template from #21 to improve clarity <3

### CODE REVIEW

File: rfcs/ui-central-dogma.md [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+This framing allows us to break the UI problem space down into many smaller, loosely coupled parts, and build our UI behaviors through **composition**.
+
+Each of these **UI building blocks** works together, answering the following questions:
+
+1. **Input dispatching**
+    1. How do we read input more than once per frame?
+    2. How do we transform input streams into actions?
+    3. How do we unify multiple input methods to the same actions?
+    4. How do we ensure actions are handed off to the correct widget?
+2. **Widget state management** ([RFC: System-driven polling architecture](https://github.com/bevyengine/rfcs/pull/11))
```
Comment (alice-i-cecile) [inline rfcs/ui-central-dogma.md:71]:
```suggestion
2. **Widget state management** ([Discussion: UI as a pure function of state](https://github.com/bevyengine/bevy/discussions/1996#discussioncomment-654221)m [RFC: System-driven polling](https://github.com/bevyengine/rfcs/pull/11))
```

**Comment by B-Reif** [comment]
Here's a little sketch of something redux/FRP like implemented with traits:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=c68897abd01df00dc38a6c9153052adb

**Comment by B-Reif** [comment]
To expand on "external tools":
> Data-driven
  ...
  How do we interface with external tools, if at all? A few developer tools to consider:

- An inspection tool, like a browser's dev tools or the [React devtools](https://chrome.google.com/webstore/detail/react-developer-tools/fmkadmapgofadopljbjfkapdkoienihi?hl=en), can help for inspecting the current state
- A changelog, like the [Redux devtools](https://chrome.google.com/webstore/detail/redux-devtools/lmhkpmbekcpmknklioeibfkpmmfibljd?hl=en), can help for inspecting app state and action history.
- A "headless" mode can help for testing without rendering.

It's worth considering these kinds of tools when designing the API. For instance, these kinds of tools usually benefit from serializable state.

**Comment by alice-i-cecile** [comment]
Two more examples I'd like to add:

1. A sudoku grid: lots of subtle complexity there. Was completely infeasible under the current UI; see: https://github.com/bevyengine/bevy/discussions/2235.
2. A dynamically generated flow-chart. Very very useful for our editor.

File: rfcs/ui-central-dogma.md [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+These paper prototypes will not actually *work*, and are intended to be composed entirely of end-user code and/or data.
+
+We've established a common set of **benchmark prototypes** below for designers to target.
+These are designed to capture a wide-range of required behavior and use cases,
+and range from very simple to frustratingly complex.
+
+Each example's unique challenges are defined, and then a detailed list of requirements is given.
+Below that, we collect examples of alternative paper prototypes, labelled with the RFCs / approaches used.
+
+TODO: add more benchmark prototypes from [#1974](https://github.com/bevyengine/bevy/issues/1974).
```
Comment (DavidVonDerau) [inline rfcs/ui-central-dogma.md:137]: https://eugenkiss.github.io/7guis/tasks/

**Pull request closed without merging**